### PR TITLE
[SEGFAULT]: root_node nullptr check

### PR DESCRIPTION
**Pull request by gaurav1086**
Accessing root_node after it may possibly still be nullptr.

### CODE REVIEW

**Review by mihaimaruseac — APPROVED** [review]

**Comment by rthadur** [comment]
@gaurav1086 there are some sanity checks failing , can you please take a look.
Here is the link:https://source.cloud.google.com/results/invocations/b15d98de-a35a-4020-8fdf-3d3978f2c401/log

**Comment by gaurav1086** [comment]
@rthadur , thanks a lot for the review. 
The two tests that are failing are: 
do_pip_smoke_test: Pip Smoke Test: Checking py_test dependencies exist in pip package
do_bazel_nobuild: bazel nobuild
They do not seem to be related to my change. I will run the tests again to see if they prevail.

**Comment by mihaimaruseac** [comment]
The sanity failures are due to https://support.sonatype.com/hc/en-us/articles/360041287334 

This is a bazel failure, hoping it will be fixed soon

**Comment by gaurav1086** [comment]
@mihaimaruseac thanks for the clarification.

**Review by mihaimaruseac — DISMISSED** [review]

**Review by sanjoy — CHANGES_REQUESTED** [review]
I don't think you need to return early, `LOG(FATAL) << msg` crashes the program.

**Comment by gaurav1086** [comment]
Yes. You are right. In file: tensorflow/core/platform/default/logging.cc:237, it calls abort(), when the LOG_LEVEL is FATAL,

 if (severity_ == FATAL) {
    abort();
}

However, I think having this change wouldn't hurt as being good coding practice.

**Comment by sanjoy** [comment]
> However, I think having this change wouldn't hurt as being good coding practice.

Can you share why you think this is good coding practice?

**Comment by gaurav1086** [comment]
@sanjoy , If someone in the future by mistake removes the "LOG(FATAL) << msg" statement or changes the log level from FATAL to something else, it might crash the code by accessing the root_node(nullptr). So good to have these additional checks (without undermining performance). However it's obviously a choice.

**Comment by sanjoy** [comment]
> @sanjoy , If someone in the future by mistake removes the "LOG(FATAL) << msg" statement or changes the log level from FATAL to something else, it might crash the code by accessing the root_node(nullptr). So good to have these additional checks (without undermining performance). However it's obviously a choice.

But there are cons to adding the `return false;` as well.  For instance if someone does not know `LOG(FATAL)` crashes , they might assume that `ParseInstructionList` legitimately returns false if `root_node` is `nullptr`.

IMO the right fix here is to add a clear comment that `LOG(FATAL)` crashes so the immediately following deref of `root_node` is OK.  What do you think?

**Comment by gaurav1086** [comment]
@sanjoy , that makes sense. Imho we should both write the comment as well as add 'return false' statement.

**Comment by sanjoy** [comment]
> @sanjoy , that makes sense. Imho we should both write the comment as well as add 'return false' statement.

I don't think we should also add the `return false;` (which will be dead code).

I'm not convinced by your argument that if someone hypothetically removes / changes `LOG(FATAL)` in the future *then* the `return false;` will help us avoid a crash.  I don't think code needs to be hardened against arbitrary mutations around it :), it will be the responsibility for whoever changed the `LOG(FATAL)` to also make sure we do the right thing for a null `root_node`.  For instance, in the future someone could remove the entire `if (root_node == nullptr) {` clause, but I don't think it is productive to modify the rest of the code so that it keeps working with that mutation.

**Comment by gaurav1086** [comment]
Done.

File: tensorflow/compiler/xla/service/hlo_parser.cc [hunk 1]
```diff
@@ -636,7 +636,7 @@ bool HloParserImpl::ParseInstructionList(HloComputation** computation,
     // the pool, which should not happen.
     if (root_node == nullptr) {
       LOG(FATAL) << "instruction " << root_name
-                 << " was marked as ROOT but the parser has not seen it before";
+                 << " was marked as ROOT but the parser has not seen it before"; // abort()
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/hlo_parser.cc]: Sorry for being so pedantic, but can you please use a full english sentence here?  `// abort()` reads like code that was accidentally commented out.

Comment (gaurav1086) [reply tensorflow/compiler/xla/service/hlo_parser.cc]: Sorry, please check now. Thanks.

File: tensorflow/compiler/xla/service/hlo_parser.cc [hunk 2]
```diff
@@ -636,7 +636,7 @@ bool HloParserImpl::ParseInstructionList(HloComputation** computation,
     // the pool, which should not happen.
     if (root_node == nullptr) {
       LOG(FATAL) << "instruction " << root_name
-                 << " was marked as ROOT but the parser has not seen it before";
+                 << " was marked as ROOT but the parser has not seen it before"; // LOG(FATAL) crashes the program by calling abort()
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/hlo_parser.cc]: Final comment: please end the sentence with period.

Comment (gaurav1086) [reply tensorflow/compiler/xla/service/hlo_parser.cc]: Added. Please check now. Thanks.

**Review by sanjoy — DISMISSED** [review]

File: tensorflow/compiler/xla/service/hlo_parser.cc [hunk 3]
```diff
@@ -636,7 +636,7 @@ bool HloParserImpl::ParseInstructionList(HloComputation** computation,
     // the pool, which should not happen.
     if (root_node == nullptr) {
       LOG(FATAL) << "instruction " << root_name
-                 << " was marked as ROOT but the parser has not seen it before";
+                 << " was marked as ROOT but the parser has not seen it before"; // LOG(FATAL) crashes the program by calling abort().
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/service/hlo_parser.cc]: `Lines should be <= 80 columns wide  [whitespace/line_length]`

Please add the comment above the `LOG(FATAL) <<` line

Comment (gaurav1086) [reply tensorflow/compiler/xla/service/hlo_parser.cc]: Added, please check. Thanks.

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**